### PR TITLE
Split authentication as an app into AppsTransport

### LINKED ISSUE #6: Provide a transport that allows authenticating against the App-level endpoints
Per the GitHub [docs](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#jwt-payload), there are some endpoints that are only available when authenticating as an app, as opposed to authenticating as an installation (specifically, the `/app` and `/app/installations` endpoints). It would be nice if the library had a way to create a client that would authenticate using that JWT token.

Fortunately, most of the code is already there - it's basically just constructing the bearer token and adding it to the `Authorization` header. It simply needs a different transport implementation. I started prototyping this, but realized it would end up duplicating the code, and thought I'd open an issue to see if you had a preferred way of doing it. Should be a pretty easy addition.

### PR DESCRIPTION
**Pull request by bradleyfalzon**
Splits authentication as a GitHub App into an `AppsTransport` which implements `http.RoundTripper`. Note this is a WIP, and requires documentation and closer review for cleanup.

For review by @itay.

Resolves #6.

### CODE REVIEW

**Comment by bradleyfalzon** [comment]
Great, gimmie 24-48 hours to clean this up, I've ran it against my integration tests in GopherCI and everything passes, so technically it's a lgtm from me.

**Comment by itay** [comment]
@bradleyfalzon thought I'd check in and see if you had a chance to get this merged. Together with https://github.com/google/go-github/pull/769, I'm trying to get this all working :)

**Comment by bradleyfalzon** [comment]
Yeah sorry mate, just need a little more time due to personal commitments, use this branch for the moment if you like. I don’t see any breaking changes will be needed, just docs and unit tests.

**Comment by itay** [comment]
Thanks! One quick note, while trying to use it, I realized there is no convenience helper for the `AppsTransport` that takes in a PEM file path rather than `[]byte` - not a deal breaker, but thought it might be good for consistency.

**Comment by bradleyfalzon** [comment]
Good feedback, I'll add it.

**Comment by shurcooL** [comment]
> Together with google/go-github#769, I'm trying to get this all working :)

There has been progress on the go-github side. :)

**Comment by bradleyfalzon** [comment]
Hi @itay, when you have a moment, I'd appreciate a second set of eyes before I squash and merge.

File: transport.go [hunk 1]
```diff
@@ -79,9 +75,9 @@ func New(tr http.RoundTripper, integrationID, installationID int, privateKey []b
 		mu:             &sync.Mutex{},
 	}
 	var err error
-	t.key, err = jwt.ParseRSAPrivateKeyFromPEM(privateKey)
+	t.appsTransport, err = NewAppsTransport(t.tr, t.integrationID, privateKey)
```
Comment (itay) [inline transport.go]: Why not pass down the `BaseURL` and `Client` here, rather than setting them on each call?

**Comment by itay** [comment]
Overall the change looks fine to me, and it does work in my particular use case. I think the only thing that's a bit weird is not passing down some of the parameters at construction time, but that's kind of a detail.

File: transport.go (hunk 1 above)
Comment (bradleyfalzon) [reply transport.go]: It's only useful for people wanting to write unit tests, or if they want to provide a different client, I would expect most people would be fine with the defaults. Making it functional args would be an option, but not worth the hassle.

**Comment by bradleyfalzon** [comment]
Thanks, I've replied, if you're happy with that response or want to challenge it let me know. I'd rather not make a breaking change to clean it up, so just want to be 100% before I merge.

**Comment by itay** [comment]
LGTM - thank you for turning it around so quickly.

File: appsTransport.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+// AppsTransport provides a http.RoundTripper by wrapping an existing
+// http.RoundTripper and provides GitHub Apps authentication as a
+// GitHub App.
+//
+// Client can also be overwritten, and is useful to change to one which
+// provides retry logic if you do experience retryable errors.
+//
+// See https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/
+type AppsTransport struct {
+	BaseURL       string            // baseURL is the scheme and host for GitHub API, defaults to https://api.github.com
```
Comment (shurcooL) [inline appsTransport.go]: Minor issue, comment incorrectly refers to the field as `baseURL` but it's actually named `BaseURL`.

Comment (bradleyfalzon) [reply appsTransport.go]: Thanks @shurcooL and yes it is, I've fixed in both places in 3f4e9b1898e6b65e31827d5dccb97d91d59feaa3

**Pull request merged**